### PR TITLE
test(tsm1): TagKeys/TagValues stops scanning when context is cancelled

### PR DESCRIPTION
**Pull request by jlapacik**
Closes #

Describe your proposed changes here.

- [ ] [CHANGELOG.md](https://github.com/influxdata/influxdb/blob/master/CHANGELOG.md) updated with a link to the PR (not the Issue)
- [ ] [Well-formatted commit messages](https://www.conventionalcommits.org/en/v1.0.0-beta.3/)
- [ ] Rebased/mergeable
- [ ] Tests pass
- [ ] http/swagger.yml updated (if modified Go structs or API)
- [ ] Documentation updated or issue created (provide link to issue/pr)
- [ ] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed)

### CODE REVIEW

File: tsdb/tsm1/engine_schema.go [hunk 1]
```diff
@@ -141,17 +163,34 @@ func (e *Engine) tagValuesPredicate(ctx context.Context, orgBucket, tagKeyBytes
...
 		tags   models.Tags
 		keybuf []byte
 		sfkey  []byte
-		stats  cursors.CursorStats
 	)
 
 	for i := range keys {
+		// to keep cache scans fast,
+		// check context every 64 iteratons.
+		if i%64 == 0 {
```
Comment (stuartcarnie) [inline tsdb/tsm1/engine_schema.go]: Can we make this an unexported const and update the references

File: tsdb/tsm1/engine_schema.go [hunk 2]
```diff
@@ -175,6 +214,14 @@ func (e *Engine) tagValuesPredicate(ctx context.Context, orgBucket, tagKeyBytes
 		}
 
 		for _, iter := range iters {
+			// Always check context before performing a tsm file scan
```
Comment (stuartcarnie) [inline tsdb/tsm1/engine_schema.go]: I'm don't think we need to check it for every iterator seek, as this operation performance a binary search, so it should execute in O(logN) time. I think every 64 iterations of the outer loop should suffice.

File: tsdb/tsm1/engine_schema.go [hunk 3]
```diff
@@ -368,6 +461,14 @@ func (e *Engine) tagKeysPredicate(ctx context.Context, orgBucket []byte, start,
...
```
Comment (stuartcarnie) [inline tsdb/tsm1/engine_schema.go]: Same here, given seek is binary search, checking every 64 keys of the outer loop should suffice.

File: tsdb/tsm1/engine_schema.go [hunk 4]
```diff
@@ -141,17 +165,33 @@ func (e *Engine) tagValuesPredicate(ctx context.Context, orgBucket, tagKeyBytes
 		return true
 	})
 
+	var stats cursors.CursorStats
+
+	if canceled {
+		stats = statsFromIters(stats, iters)
+		return cursors.NewStringSliceIteratorWithStats(nil, stats), ctx.Err()
```
Comment (e-dard) [inline tsdb/tsm1/engine_schema.go]: I suppose this is getting a little philosophical now, but is this an error or not? `tagValuesPredicate` didn't hit any particular error conditions, it was just told to finish up early (due to deadline or explicit cancellation). It's clear here that you want to inform the caller of the (partial) work that was done. 

To me, this feels a little bit like the API should not be returning an error, but inside of the result value (the iterator), we should actually have a `Err()` method, which callers should check. That would return an error if `tagValuesPredicate` was told to terminate early.

To me, if a method returns a result and an error, and the error is `non-nil`, then the result is always undefined or useless.

File: tsdb/tsm1/engine_schema.go [hunk 5]
```diff
@@ -14,6 +14,8 @@ import (
 	"github.com/influxdata/influxql"
 )
 
+const contextPeriod = 64
```
[the remark this answers is not in the document]
Comment (jlapacik) [reply tsdb/tsm1/engine_schema.go]: Sure thing. Yes keys scans.

File: tsdb/tsm1/engine_schema.go [hunk 6]
```diff
@@ -113,7 +129,7 @@ func (e *Engine) tagValuesPredicate(ctx context.Context, orgBucket, tagKeyBytes
 
 	keys, err := e.findCandidateKeys(ctx, orgBucket, predicate)
 	if err != nil {
-		return nil, err
+		return cursors.EmptyStringIterator, err
```
[the remark this answers is not in the document]
Comment (jlapacik) [reply tsdb/tsm1/engine_schema.go]: Initially I was not returning an error when the context was canceled. Instead I just returned the partial work, i.e the number of bytes/values scanned up until that point. However after looking at the godoc for `context.Done()` as well as other examples, it seemed an error should be returned for a canceled context.

I decided to split the difference, returning a non-nil error in the event the context is canceled while at the same time returning the partial work up until that point. Note there is precedent for such return semantics. As identified in a slack conversations, the stdlib's `io.Reader` is one such example.

Also I have verified that no callers of TagKeys/TagValues rely on a `nil` string iterator in the event of a non-nil error.

File: tsdb/tsm1/engine_schema.go (hunk 4 above)
Comment (jlapacik) [reply tsdb/tsm1/engine_schema.go]: Yes I think canceling the context should result in an error. Were we to not treat it as an error, we would run into inconsistencies higher up the stack where a canceled context is treated as an error. Also as mentioned above (from a slack conversation) there is a precedent, set by the stdlib, for returning both a non-nil error as well as non-zero value result.

Having said that I will make it explicit in the godoc for TagKeys/TagValues that this is the case.

**Comment by jlapacik** [comment]
alchemy please merge

**Pull request merged**